### PR TITLE
Fix furnace fuel being stupid

### PR DESCRIPTION
**Pull request by Shadows-of-Fire**
Furnace fuel is now accessable statically in ForgeHooks, and now caches the vanilla burn map.  It is updated when tags are reloaded so that tag-based burns are updated.

Client data events:

TagsUpdatedEvent is now provided for the client.  It has a NetworkTagManager context and is fired when the sync packet is received.

RecipesUpdatedEvent has had a List<IRecipe<?>> context added to it.

### CODE REVIEW

File: src/main/java/net/minecraftforge/common/ForgeMod.java [hunk 1]
```diff
@@ -141,6 +144,22 @@ public void serverStarting(FMLServerStartingEvent evt)
     {
         new ForgeCommand(evt.getCommandDispatcher());
         ConfigCommand.register(evt.getCommandDispatcher());
+        evt.getServer().getResourceManager().addReloadListener(new ReloadListener<Object>()
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/common/ForgeMod.java]: Why is this needed if we get the reload event?

Comment (Shadows-of-Fire) [reply src/main/java/net/minecraftforge/common/ForgeMod.java]: The TagsUpdatedEvent is only posted on the client when the sync packet is received.  The server needs to update on data reload.

Comment (LexManos) [reply src/main/java/net/minecraftforge/common/ForgeMod.java]: Seems like a lack in the update event.
I'd have to look into it to be sure, but I think reload listener should be fired on both client and server in this context.

**Comment by Shadows-of-Fire** [comment]
The events are probably more aptly named (Tags/Recipes)SyncedEvent instead of Updated, due to how they work.
In regards to the last comment above, data reloads are never fired on the client when connected to a dedicated server.

File: patches/minecraft/net/minecraft/tileentity/AbstractFurnaceTileEntity.java.patch [hunk 2]
```diff
@@ -1,5 +1,14 @@
 --- a/net/minecraft/tileentity/AbstractFurnaceTileEntity.java
 +++ b/net/minecraft/tileentity/AbstractFurnaceTileEntity.java
+@@ -84,7 +84,7 @@
+       super(p_i49964_1_);
+       this.field_214014_c = p_i49964_2_;
+    }
+-
++   @Deprecated //Forge - get burn times by calling ForgeHooks#getBurnTime(ItemStack)
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/tileentity/AbstractFurnaceTileEntity.java.patch:7]: keep the empty line

File: patches/minecraft/net/minecraft/tileentity/AbstractFurnaceTileEntity.java.patch [hunk 3]
```diff
@@ -71,21 +80,20 @@
           Item item = p_213997_1_.func_77973_b();
 -         return func_214001_f().getOrDefault(item, 0);
 +         int ret = p_213997_1_.getBurnTime();
-+         return net.minecraftforge.event.ForgeEventFactory.getItemBurnTime(p_213997_1_, ret == -1 ? func_214001_f().getOrDefault(item, 0) : ret);
++         return net.minecraftforge.common.ForgeHooks.getBurnTime(p_213997_1_);
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/tileentity/AbstractFurnaceTileEntity.java.patch:81]: extra ret that is unused

File: src/main/java/net/minecraftforge/client/ForgeHooksClient.java [hunk 4]
```diff
@@ -1010,9 +1011,9 @@ public static boolean onGuiCharTypedPost(Screen guiScreen, char codePoint, int m
         return MinecraftForge.EVENT_BUS.post(event);
     }
 
-    public static void onRecipesUpdated()
+    public static void onRecipesUpdated(List<IRecipe<?>> recipes)
     {
-        Event event = new RecipesUpdatedEvent();
+        Event event = new RecipesUpdatedEvent(recipes);
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/client/ForgeHooksClient.java]: This is a binary breaking change, and also you should probably pass in the recipe book/recipe manager and not the raw list of recipes... As the manager has more information. And should have access to this list right?

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 5]
```diff
@@ -1179,4 +1183,30 @@ public static boolean canEntityDestroy(World world, BlockPos pos, LivingEntity e
...
+        }
+        else
+        {
+            Item item = stack.getItem();
+            int ret = stack.getBurnTime();
+            return ForgeEventFactory.getItemBurnTime(stack, ret == -1 ? VANILLA_BURNS.getOrDefault(item.delegate, 0) : ret);
+        }
+    }
+
+    public static void updateBurns()
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: Sync this, and do we need to keep a server/client list in case tags are not synced correctly?

**Review by LexManos — CHANGES_REQUESTED** [review]

File: src/main/java/net/minecraftforge/client/ForgeHooksClient.java (hunk 4 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/minecraftforge/client/ForgeHooksClient.java]: The manager does not have access to the passed list only the total recipes available on the client.  And while technically breaking, unless someone is manually firing the event (which nobody should ever be, as recipe loading is internal) this will not be a problem.

It could stand to access the recipe manager, so that can be added, but both may want to be present.

Comment (LexManos) [reply src/main/java/net/minecraftforge/client/ForgeHooksClient.java]: Is this recipe list just a subset? How does this system work?
Can the server send remove, as well as add?
It's been a while since I've looked at recipes, but arnt they a full Forge registry and thus synced at connect? At-least ID wise.

Comment (Shadows-of-Fire) [reply src/main/java/net/minecraftforge/client/ForgeHooksClient.java]: We removed the recipe registry when the manager was introduced in 1.13, so recipes no longer have (numeric) ids and can be synced whenever desired.  However it looks like the vanilla expectancy is that the packet always sends the entire recipe list to the client, so the collection can be dropped in favor of the entire manager.

As for when the packet is sent, it is sent on login and also re-sent whenever /reload is executed.

File: src/main/java/net/minecraftforge/common/ForgeHooks.java (hunk 5 above)
Comment (Shadows-of-Fire) [reply src/main/java/net/minecraftforge/common/ForgeHooks.java]: I'm not sure if it makes any sense to be synchronized, as it should only be executed on the main thread (and being synchronized would not prevent CME's as the map is not synchronized).

Comment (LexManos) [reply src/main/java/net/minecraftforge/common/ForgeHooks.java]: The only read is not in a loop so there wouldnt be any CMEs EXCEPT this function that does a ton of writes.
It's just a safty thing on a non-critical section.

**Comment by Shadows-of-Fire** [comment]
All concerns addressed
